_episodes/l6-06-Best_practices.md

```diff
@@ -0,0 +1,67 @@
+    - Avoid writing tests that depend on the state or side effects of other tests. Each test case should be independent and self-contained.
+    - Use test fixtures to encapsulate common setup and teardown logic, promoting code reusability and reducing code duplication.
+    - In general follow the rule **ONE ASSERT PER TEST** as much as possible.
+
+3. **Write Clear and Readable Tests**:
+    - Use comments to explain the purpose and expected behaviour of each test case, especially when dealing with complex or edge cases. A comment would enhance the readability of the test.
+    - Break down complex test scenarios into smaller, manageable assertions. Each assertion should test a single condition.
+
+4. **Choose Appropriate Assertions**:
+    - Select the most appropriate assertion macros provided by Google Test that closely match the behaviour being tested.
+    - Use specific assertions (e.g., ASSERT_EQ, ASSERT_TRUE, ASSERT_FALSE) instead of general assertions (e.g., ASSERT or EXPECT) to improve test failure diagnosis.
+
+5. **Test Coverage and Code Review**:
+    - Aim for high code coverage by ensuring that the tests exercise different code paths, including error handling, edge cases, and boundary conditions.
+    - Regularly review your test suite to identify gaps in test coverage and update it accordingly. Periodically review and remove redundant or obsolete tests.
+    - Use code coverage analysis tools to measure the effectiveness of your tests and identify areas that need additional testing.
+
+6. **Maintain Test Independence**:
+    - Avoid sharing mutable state between tests. Each test case should start with a clean and consistent state to ensure test independence.
+    - Reset any modified state or resources in the test teardown phase to leave the environment in a predictable state for subsequent tests.
+    - Be cautious when using global variables or static state in tests, as they can introduce unwanted dependencies and make tests more fragile.
+
+7. **Mocking and Test Doubles**:
+    - Use mocking frameworks, such as Google Mock, to create test doubles (mock objects) for dependencies that need to be isolated or controlled during testing.
+
+8. **Continuous Integration and Test Execution**:
+    - Integrate your unit tests into your continuous integration (CI) process to ensure that tests are automatically executed with each code change.
+    - Aim for fast and deterministic tests by minimising external dependencies, reducing I/O operations, and avoiding non-deterministic behaviours.
+
+9. **Test Failure Investigation**:
+    - When a test fails, investigate and diagnose the failure by examining the failure message, log output, and any relevant debug information.
+
+10. **Ensure Repeatability**:
+    - Ensure that the unit tests produce the same results when run at different times, different machines etc.
+
+### Summary
+
+In this chapter, we described some of the best practices that should be followed while writing unit tests.
+
```

Comment:
We could maybe also talk about in-source test code vs a dedicated test code directory. Advantage of in-source tests are that they are physically located close to the actual code that they will test,m which (hypothetically) makes developers more likely to update tests when they update code. What do you think?

Reply:
Would this increase the size of the executables/binaries?

I see the usefulness of keeping the tests alongside the code they are testing, but mixing functionality with development-only tooling feels odd. I guess there are pros and cons and might boil down to a question of preference.

Reply:
I think the idea is to still build them separately, but the act of placing test files next to source files (n.b. different files) is like a gentle poke to remind devs to update tests when they update code. Not sure if it actually works, though...

Reply:
I think we can probably skip this discussion for a future time as there's no obvious agreement on what's best practice on that front.